tsdb/index/postings.go

```diff
@@ -292,30 +293,76 @@ func (p *MemPostings) EnsureOrder(numberOfConcurrentProcesses int) {
 func (p *MemPostings) Delete(deleted map[storage.SeriesRef]struct{}, affected map[labels.Label]struct{}) {
 	p.mtx.Lock()
 	defer p.mtx.Unlock()
+	if len(p.m) == 0 || len(deleted) == 0 {
+		return
+	}
```

Comment:
Not sure if it would make any difference but up to this point you only needed a read lock

Reply:
If your MemPostings are empty, nothing makes a difference 😂

Reply:
But this raises an issue, let me write an additional test to make sure that we don't deadlock if we remove all postings.

Reply:
Added some more tests: https://github.com/prometheus/prometheus/pull/14975/commits/ada8a6ef10c37ec0ea37b2e0c21e4ec2187a6fa8

I originally added this shortcut because if MemPostings are empty, we initialize the `deleteLabelNames` channel with 0 capacity, but we still process `allPostingsKey` and try to delete it. Now that case is tested.